src/sync.cpp

```diff
@@ -172,6 +172,7 @@ static void push_lock(MutexType* c, const CLockLocation& locklocation)
             // same thread as that results in an undefined behavior.
             auto lock_stack_copy = lock_stack;
             lock_stack.pop_back();
+            lock.unlock();
```

Comment:
In commit "Release `LockData::dd_mutex` before calling `*_detected` functions" (08cfe273ddf9bd7640456865926ec9d94b363520)

Would be good to have a comment saying why this is necessary. If I understand correctly it could say something like // Release dd_mutex before calling double_lock_detected to avoid deadlocks, so no other lock is ever acquired after dd_mutex, and lock order is consistent.